### PR TITLE
Piggy back on request promise for 2.5 seconds

### LINKED ISSUE #1640: User settings request is duplicated multiple times and request made frequently while navigating app
image

### PR DESCRIPTION
**Pull request by ghost**
# Description

An optimization workaround that prevents weaver applications from making duplicate requests from models and repos. This is common when directives and controllers are not carefully scoped.

There is a hardcoded 2500 millisecond timeout for duplicate requests to share (piggy back) a request promise made before. This may be best to be in appConfig.js.

## Type of change

- [x] Optimization workaround

# How Has This Been Tested?

Please check box below when reviewed.

- [x] [Vireo](https://github.com/TexasDigitalLibrary/Vireo)

### CODE REVIEW

**Review by rmathew1011 — APPROVED** [review]

File: app/model/abstractModel.js [hunk 1]
```diff
@@ -59,9 +61,20 @@ core.factory("AbstractModel", function ($injector, $rootScope, $q, ModelCache, M
                 }
 
                 if (fetch) {
-                    WsApi.fetch(mapping.instantiate).then(function (res) {
+                    if (fetchingPromise === undefined) {
+                        console.log('fetching', mapping.instantiate);
```
Comment (kaladay) [inline app/model/abstractModel.js]: Could we avoid console logging on every fetch?
I am referring to this console log ferenced and the others added.

**Review by kaladay — CHANGES_REQUESTED** [review]

File: app/repo/abstractRepo.js [hunk 2]
```diff
@@ -383,10 +385,23 @@ core.service("AbstractRepo", function ($q, $rootScope, $timeout, ApiResponseActi
 
         var fetch = function () {
             if (abstractRepo.mapping.all !== undefined) {
-                WsApi.fetch(abstractRepo.mapping.all).then(function (res) {
+                if (fetchingAllPromise === undefined) {
+                    console.log('fetching', abstractRepo.mapping.all);
+                    fetchingAllPromise = WsApi.fetch(abstractRepo.mapping.all);
+                } else  {
```
Comment (kaladay) [inline app/repo/abstractRepo.js]: Please remove extra space after `else`.

**Review by kaladay — COMMENTED** [review]
What happens when two or more requests are handling the same request but one fails before the other completes?

It looks like the promise is being removed and a reject is being returned.

Could the rejection affect the other request?
Has this been tested?

**Comment by ghost** [comment]
> What happens when two or more requests are handling the same request but one fails before the other completes?

_The first request fail/pass state will determine response for duplicates within 2500 milliseconds._

> It looks like the promise is being removed and a reject is being returned.

_The error is being caught and rejected appropriately from the Web Service API. And effectively removing the definition from the promise._

> Could the rejection affect the other request? Has this been tested?

_For the duration of the defer, the promise will be handled if subscribed to already. Yes, it has been tested in the Vireo admin submission list view._

**Comment by kaladay** [comment]
> The subsequent requests will have a failed state if a duplicate request is issued within approximately 2.5 seconds.

This is a huge concern of mine.

Moving on, analytics shows some odd behavior when doing some testing.

Visually speaking when using the changes here the page appears to load faster.
When viewing the network log the traffic appears to block less.

When using the performance tester and even Lighthouse, the weaver changes appears to take longer and be slower. Looking at the data I feel inclined to consider Lighthouse unreliable.

Note: These tests are against the Vireo branch without the associated Vireo PR and without these changes verses the branch with the associated Vireo PR and these changes. Some of the differences observed here may be related to the Many to Many Lazy change in Vireo.

**Comment by wwelling** [comment]
Can you perform each test 5 times and provide the average as well?

**Comment by kaladay** [comment]
> Can you perform each test 5 times and provide the average as well?

No, you can go ahead and do that if you want to.


For better analysis, here are the analysis comparing with/without vireo changes.

### XHR Traffic With These Changes but Without Vireo Changes
with_weaver_changes-without_vireo_changes-xhr_traffic

### XHR Traffic Without These Changes but With Vireo Changes
without_weaver_changes-with_vireo_changes-xhr_traffic

### Performance With These Changes but Without Vireo Changes
with_weaver_changes-without_vireo_changes-performance

### Performance Without These Changes but With Vireo Changes
without_weaver_changes-with_vireo_changes-performance

### Lighthouse With These Changes but Without Vireo Changes
with_weaver_changes-without_vireo_changes-lighthouse

### Lighthouse Without These Changes but With Vireo Changes
without_weaver_changes-with_vireo_changes-lighthouse

**Comment by wwelling** [comment]
Please remove all local machine measurements from the review comments please.

**Comment by wwelling** [comment]
> Some of the differences observed here may be related to the Many to Many Lazy change in Vireo.

Can you perform your local measured tests with or without the @wvr/core dependency change?

**Comment by ghost** [comment]
> This is a huge concern of mine.

What is the concern? Duplicate requests are not appropriate by any means by a client application. We are not long polling; we have adopted web socket communications.

**Comment by wwelling** [comment]
> > The subsequent requests will have a failed state if a duplicate request is issued within approximately 2.5 seconds.
> 
> This is a huge concern of mine.

Is the concern my poor grammar? I must have to read and write more.

**Comment by ghost** [comment]
@kaladay I am willing to put a configuration in appConfig.js to enable/disable and another for the wait for duplicate requests. This will afford us to disable in Vireo and fix the scoping problems that are causing the duplicate requests. Other weaver ui apps are far less likely to be making duplicate requests.

**Review by kaladay — APPROVED** [review]

**Comment by ghost** [comment]
@jeremythuff @jcreel is this safe to merge and cut a rc?

**Pull request merged**